Flaky test ProgrammingExerciseBitbucketBambooIntegrationTest > testExportCourseCannotExportSingleParticipationGitApiException

### Describe the bug

The test ProgrammingExerciseBitbucketBambooIntegrationTest > testExportCourseCannotExportSingleParticipationGitApiException is flaky, in particular on Github actions. We should fix it so that all tests pass again

### To Reproduce

Run the tests on Github action as part of any PR

### Expected behavior

The test should pass reliably

### Screenshots

_No response_

### What browsers are you seeing the problem on?

Other (specify in "Additional context")

### Additional context

_No response_

### Relevant log output

```shell
2021-07-15T20:35:42.1897812Z ProgrammingExerciseBitbucketBambooIntegrationTest > testExportCourseCannotExportSingleParticipationGitApiException() STANDARD_OUT
2021-07-15T20:35:42.1904911Z     2021-07-15 20:35:42.110  INFO 1935 --- [    Test worker] d.t.i.w.a.s.CourseExamExportService      : Exporting course exercises for course 210 and title Course title b23f8b37-8507-4382-86b6-64629d6d2ad0
2021-07-15T20:35:42.1908957Z     2021-07-15 20:35:42.110  INFO 1935 --- [    Test worker] d.t.i.w.a.s.CourseExamExportService      : Exporting exercise adbef620-d539-4bf1-b01d-fb08cfb0ca74 with id 316 
2021-07-15T20:35:42.1913259Z     2021-07-15 20:35:42.113  INFO 1935 --- [    Test worker] w.a.s.p.ProgrammingExerciseExportService : Exporting programming exercise 316 with title adbef620-d539-4bf1-b01d-fb08cfb0ca74
2021-07-15T20:35:42.1918633Z     2021-07-15 20:35:42.116  INFO 1935 --- [    Test worker] w.a.s.p.ProgrammingExerciseExportService : Request to export the repositories of programming exercise 316 with title 'adbef620-d539-4bf1-b01d-fb08cfb0ca74' of the following students or teams: student1, student2
2021-07-15T20:35:42.3895344Z     2021-07-15 20:35:42.311 DEBUG 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Fetch /tmp/studentRepo16783745728091228925
2021-07-15T20:35:42.3897540Z     2021-07-15 20:35:42.312  WARN 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot set remoteUrl because it is null!
2021-07-15T20:35:42.3900068Z     2021-07-15 20:35:42.322 ERROR 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot hard reset the repo /tmp/studentRepo16783745728091228925 to origin/HEAD because it is empty.
2021-07-15T20:35:42.3902593Z     2021-07-15 20:35:42.322 DEBUG 1935 --- [    Test worker] w.a.s.p.ProgrammingExerciseExportService : Create temporary zip file for repository /tmp/studentRepo16783745728091228925
2021-07-15T20:35:42.3905156Z     2021-07-15 20:35:42.331 ERROR 1935 --- [    Test worker] w.a.s.p.ProgrammingExerciseExportService : Failed to create zip for participation id 244 with exercise id 316 because of the following exception 
2021-07-15T20:35:42.3906439Z 
2021-07-15T20:35:42.3907672Z     org.eclipse.jgit.api.errors.InvalidRemoteException: InvalidRemoteException
2021-07-15T20:35:42.3910412Z     	at de.tum.in.www1.artemis.service.connectors.GitService.getOrCheckoutRepository(GitService.java:346)
2021-07-15T20:35:42.3913770Z     	at de.tum.in.www1.artemis.service.connectors.GitService.getOrCheckoutRepository(GitService.java:285)
2021-07-15T20:35:42.3924571Z     	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)
2021-07-15T20:35:42.3926068Z     	at java.base/java.lang.Thread.run(Thread.java:831)
2021-07-15T20:35:42.3926568Z 
2021-07-15T20:35:42.3934860Z     2021-07-15 20:35:42.335  INFO 1935 --- [    Test worker] w.a.s.p.ProgrammingExerciseExportService : Request to export instructor repository of type exercise of programming exercise ProgrammingExercise{id=316, templateRepositoryUrl='http://some.test.url/scm/SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB/shorta96272f6047ac04ceb0ae0f0aa172e2dfecat6cb-exercise.git', solutionRepositoryUrl='http://some.test.url/scm/SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB/shorta96272f6047ac04ceb0ae0f0aa172e2dfecat6cb-solution.git', templateBuildPlanId='SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB-BASE', solutionBuildPlanId='SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB-SOLUTION', publishBuildPlanUrl='null', allowOnlineEditor='null', programmingLanguage='JAVA', packageName='de.test', testCasesChanged='false'} with title 'adbef620-d539-4bf1-b01d-fb08cfb0ca74'
2021-07-15T20:35:42.3948849Z     2021-07-15 20:35:42.336 DEBUG 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Fetch /tmp/exerciseLocalRepo15227157471784351287
2021-07-15T20:35:42.3951261Z     2021-07-15 20:35:42.336  WARN 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot set remoteUrl because it is null!
2021-07-15T20:35:42.3953670Z     2021-07-15 20:35:42.342 ERROR 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot hard reset the repo /tmp/exerciseLocalRepo15227157471784351287 to origin/HEAD because it is empty.
2021-07-15T20:35:42.3964650Z     2021-07-15 20:35:42.350  INFO 1935 --- [    Test worker] w.a.s.p.ProgrammingExerciseExportService : Request to export instructor repository of type solution of programming exercise ProgrammingExercise{id=316, templateRepositoryUrl='http://some.test.url/scm/SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB/shorta96272f6047ac04ceb0ae0f0aa172e2dfecat6cb-exercise.git', solutionRepositoryUrl='http://some.test.url/scm/SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB/shorta96272f6047ac04ceb0ae0f0aa172e2dfecat6cb-solution.git', templateBuildPlanId='SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB-BASE', solutionBuildPlanId='SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB-SOLUTION', publishBuildPlanUrl='null', allowOnlineEditor='null', programmingLanguage='JAVA', packageName='de.test', testCasesChanged='false'} with title 'adbef620-d539-4bf1-b01d-fb08cfb0ca74'
2021-07-15T20:35:42.4000902Z     2021-07-15 20:35:42.352 DEBUG 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Fetch /tmp/solutionLocalRepo10574534405401785803
2021-07-15T20:35:42.4003081Z     2021-07-15 20:35:42.352  WARN 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot set remoteUrl because it is null!
2021-07-15T20:35:42.4005685Z     2021-07-15 20:35:42.359 ERROR 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot hard reset the repo /tmp/solutionLocalRepo10574534405401785803 to origin/HEAD because it is empty.
2021-07-15T20:35:42.4901080Z     2021-07-15 20:35:42.367  INFO 1935 --- [    Test worker] w.a.s.p.ProgrammingExerciseExportService : Request to export instructor repository of type tests of programming exercise ProgrammingExercise{id=316, templateRepositoryUrl='http://some.test.url/scm/SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB/shorta96272f6047ac04ceb0ae0f0aa172e2dfecat6cb-exercise.git', solutionRepositoryUrl='http://some.test.url/scm/SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB/shorta96272f6047ac04ceb0ae0f0aa172e2dfecat6cb-solution.git', templateBuildPlanId='SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB-BASE', solutionBuildPlanId='SHORTA96272F6047AC04CEB0AE0F0AA172E2DFECAT6CB-SOLUTION', publishBuildPlanUrl='null', allowOnlineEditor='null', programmingLanguage='JAVA', packageName='de.test', testCasesChanged='false'} with title 'adbef620-d539-4bf1-b01d-fb08cfb0ca74'
2021-07-15T20:35:42.4909048Z     2021-07-15 20:35:42.369 DEBUG 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Fetch /tmp/testLocalRepo15221522729483345773
2021-07-15T20:35:42.4911328Z     2021-07-15 20:35:42.369  WARN 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot set remoteUrl because it is null!
2021-07-15T20:35:42.4913846Z     2021-07-15 20:35:42.375 ERROR 1935 --- [    Test worker] d.t.i.w.a.service.connectors.GitService  : Cannot hard reset the repo /tmp/testLocalRepo15221522729483345773 to origin/HEAD because it is empty.
2021-07-15T20:35:42.4917166Z     2021-07-15 20:35:42.390  WARN 1935 --- [    Test worker] d.t.i.w.a.s.CourseExamExportService      : Skipping exam export since the course does not have any exams
2021-07-15T20:35:42.4920101Z     2021-07-15 20:35:42.399 ERROR 1935 --- [    Test worker] d.t.i.w.a.s.CourseExamExportService      : Failed to create zip file./exports/server-integration-test/courses/shorta96272f6047ac04ceb0ae0f0aa172e2dfeca-Course_title_b23f8b37-8507-4382-86b6-64629d6d2ad0-20210715-203542.zip.
2021-07-15T20:35:42.4921994Z 
2021-07-15T20:35:42.4924018Z     java.nio.file.NoSuchFileException: ./exports/server-integration-test/courses/shorta96272f6047ac04ceb0ae0f0aa172e2dfeca-Course_title_b23f8b37-8507-4382-86b6-64629d6d2ad0-20210715-203542.zip
2021-07-15T20:35:42.4926744Z     	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
2021-07-15T20:35:42.4928680Z     	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
2021-07-15T20:35:42.4931087Z     	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
2021-07-15T20:35:42.4933362Z     	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
2021-07-15T20:35:42.4935689Z     	at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)
2021-07-15T20:35:42.4937519Z     	at java.base/java.nio.file.Files.newOutputStream(Files.java:224)
2021-07-15T20:35:42.4939371Z     	at de.tum.in.www1.artemis.service.ZipFileService.createZipFile(ZipFileService.java:46)
2021-07-15T20:35:42.4941594Z     	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)
2021-07-15T20:35:42.4943026Z     	at java.base/java.lang.Thread.run(Thread.java:831)
2021-07-15T20:35:42.4943521Z 
2021-07-15T20:35:42.4945001Z     2021-07-15 20:35:42.404  INFO 1935 --- [    Test worker] d.t.i.w.a.s.CourseExamExportService      : Successfully exported course 210. The zip file is located at: Optional.empty
2021-07-15T20:35:42.7892626Z 
2021-07-15T20:35:42.7895515Z ProgrammingExerciseBitbucketBambooIntegrationTest > testExportCourseCannotExportSingleParticipationGitApiException() FAILED
2021-07-15T20:35:42.7898106Z     java.lang.AssertionError: 
2021-07-15T20:35:42.7899318Z     Expecting Optional to contain a value but it was empty.
2021-07-15T20:35:42.7905402Z         at de.tum.in.www1.artemis.programmingexercise.ProgrammingExerciseTestService.testExportCourseWithFaultyParticipationCannotGetOrCheckoutRepository(ProgrammingExerciseTestService.java:998)
2021-07-15T20:35:42.7916241Z         at de.tum.in.www1.artemis.programmingexercise.ProgrammingExerciseTestService.testExportCourseCannotExportSingleParticipationGitApiException(ProgrammingExerciseTestService.java:980)
2021-07-15T20:35:42.7929434Z         at de.tum.in.www1.artemis.programmingexercise.ProgrammingExerciseBitbucketBambooIntegrationTest.testExportCourseCannotExportSingleParticipationGitApiException(ProgrammingExerciseBitbucketBambooIntegrationTest.java:304)
2021-07-15T20:35:42.8892848Z
```

--- a/src/main/java/de/tum/in/www1/artemis/service/CourseExamExportService.java
+++ b/src/main/java/de/tum/in/www1/artemis/service/CourseExamExportService.java
@@ -432,6 +432,12 @@ public class CourseExamExportService {
      */
     private Optional<Path> createCourseZipFile(Path outputZipFile, List<Path> filesToZip, Path relativeZipPath, List<String> exportErrors) {
         try {
+            // Create the parent directories if they don't exist otherwise the zip file cannot be created.
+            Path parentDir = outputZipFile.getParent();
+            if (!Files.exists(parentDir)) {
+                Files.createDirectories(parentDir);
+            }
+
             zipFileService.createZipFile(outputZipFile, filesToZip, relativeZipPath);
             log.info("Successfully created zip file: {}", outputZipFile);
             return Optional.of(outputZipFile);

